server/src/main/java/org/opensearch/cluster/routing/FailOpenRouting.java

```diff
@@ -22,29 +22,39 @@
 import org.opensearch.transport.NodeDisconnectedException;
 import org.opensearch.transport.NodeNotConnectedException;
 
+import java.util.List;
 import java.util.Map;
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.stream.Stream;
 
 /**
- * This class contains logic to decide if retryable shard search requests can be tried on shard copies present in data
+ * This class contains logic to find next shard to retry search request in case of failure from other shard copy.
+ * This decides if retryable shard search requests can be tried on shard copies present in data
  * nodes whose attribute value weight for weighted shard routing is set to zero.
  */
 
 public class FailOpenRouting {
 
     private static final Logger logger = LogManager.getLogger(FailOpenRouting.class);
 
+    /* exception causing failure for previous shard copy */
+    private Exception exception;
+
+    private ClusterState clusterState;
+
+    public FailOpenRouting(Exception e, ClusterState clusterState) {
+        this.exception = e;
+        this.clusterState = clusterState;
+    }
+
     /**
      * *
-     * @param e exception causing failure for previous shard copy
      * @return true if exception is due to cluster availability issues
      */
-    private static boolean isInternalFailure(Exception e) {
-        return e instanceof NoShardAvailableActionException
-            || e instanceof UnavailableShardsException
-            || e instanceof NodeNotConnectedException
-            || e instanceof NodeDisconnectedException;
+    private boolean isInternalFailure() {
+        return exception instanceof NoShardAvailableActionException
+            || exception instanceof UnavailableShardsException
+            || exception instanceof NodeNotConnectedException
+            || exception instanceof NodeDisconnectedException;
```

Comment:
`exception` is always going to be `OpenSearchException` and its status will be 5XX for it to be retried. Can we use this instead of hardcoding this exceptions types ?

Reply:
yeah true, that would make this logic simpler but I am not able to find a way to get status. Do you have any pointer on this?

Reply:
We can use something like below 

```
        assert exception instanceof OpenSearchException;
         return ((OpenSearchException) exception).status().getStatus() / 100 == 5;
```

Reply:
Thanks @gbbafna. Have made the changes to check for 5xx status